### PR TITLE
Alphanumeric QR code encoding support

### PR DESCRIPTION
**Pull request by mpichette**
First pass of alphanumeric QR code support. Still need to do the following:

1. add additional tests to QRCode.DataEncoding
2. centralize some of the custom guards and import them
3. figure out how the pattern tests are done and re-create for an alphanumeric mode QR

### CODE REVIEW

File: lib/qr_code/alphanumeric_encoding.ex [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    %{qr | encoded: encoded}
+  end
+
+  def convert_chunks(codeword, acc \\ [])
+
+  def convert_chunks(<<first::utf8, second::utf8, rest::binary>>, acc)
+      when is_map_key(@mapping, first) and is_map_key(@mapping, second) do
+    chunk_value = @mapping[first] * 45 + @mapping[second]
+    #    padded_chunk = chunk_value |> Integer.to_string(2) |> String.pad_leading(11, "0")
```
Comment (mpichette) [inline lib/qr_code/alphanumeric_encoding.ex]:
```suggestion
```

File: lib/qr_code/alphanumeric_encoding.ex [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    padded_chunk = <<chunk_value::size(11)>>
+    convert_chunks(rest, [padded_chunk | acc])
+  end
+
+  def convert_chunks(<<last::utf8>>, acc) when is_map_key(@mapping, last) do
+    #    padded_value = @mapping[last] |> Integer.to_string(2) |> String.pad_leading(6, "0")
```
Comment (mpichette) [inline lib/qr_code/alphanumeric_encoding.ex]:
```suggestion
```

File: lib/qr_code/alphanumeric_encoding.ex [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    padded_value = <<@mapping[last]::size(6)>>
+    convert_chunks(<<>>, [padded_value | acc])
+  end
+
+  def convert_chunks(<<>>, acc) do
+    #    acc |> Enum.reverse() |> List.to_string()
```
Comment (mpichette) [inline lib/qr_code/alphanumeric_encoding.ex]:
```suggestion
```

File: lib/qr_code/alphanumeric_encoding.ex [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    nil
+  end
+
+  #  defp zero_pad(expected_length, chunk) when bit_size(chunk) < expected_length do
+  #    <<0::size(expected_length - bit_size(chunk))>>
+  #  end
+  #
+  #  defp zero_pad(expected_length, chunk) do
+  #    chunk
+  #  end
```
Comment (mpichette) [inline lib/qr_code/alphanumeric_encoding.ex]:
```suggestion
```

File: lib/qr_code/byte_mode.ex [hunk 5]
```diff
@@ -174,38 +172,52 @@ defmodule QRCode.ByteMode do
...
+  #
+  #  defp find_version(level, bytes) do
+  #    Enum.reduce_while(level, {:error, "Input string can't be encoded"}, fn {max, ver}, acc ->
+  #      if bytes <= max do
+  #        {:halt, {:ok, ver}}
+  #      else
+  #        {:cont, acc}
+  #      end
+  #    end)
+  #  end
```
Comment (mpichette) [inline lib/qr_code/byte_mode.ex]:
```suggestion
```

File: lib/qr_code/character_capacity.ex [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  defguard is_version(value) when is_integer(value) and value > 0 and value <= 40
+  #  @capacities (for mode <- @modes, into: %{} do
+  #                 level_map =
+  #                   for level <- @levels, into: %{} do
+  #                     converted = for {m, l} <- __MODULE__.get_level(level, mode), into: %{}, do: {l, m}
+  #                     {level, converted}
+  #                   end
+  #
+  #                 {mode, level_map}
+  #               end)
```
Comment (mpichette) [inline lib/qr_code/character_capacity.ex]:
```suggestion
```

File: lib/qr_code/error_correction.ex [hunk 7]
```diff
@@ -292,12 +292,12 @@ defmodule QRCode.ErrorCorrection do
   end
 
   defp put_info(%__MODULE__{} = ecc, version, level) do
-    {ec_codewrods_per_block, blocks_in_group1, codewords_in_group1, blocks_in_group2,
+    {ec_codewords_per_block, blocks_in_group1, codewords_in_group1, blocks_in_group2,
```
Comment (photomattmills) [inline lib/qr_code/error_correction.ex:295]: Good catch.

**Review by photomattmills — APPROVED** [review]
Looks solid.

**Pull request merged**